connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/PluginScannerTest.java

```diff
@@ -20,79 +20,114 @@
 import org.junit.Rule;
 import org.junit.Test;
 import org.junit.rules.TemporaryFolder;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
 
 import java.nio.file.Files;
 import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collection;
 import java.util.Collections;
+import java.util.HashSet;
 import java.util.List;
 import java.util.Set;
 
-import static org.junit.Assert.assertNotNull;
-import static org.junit.Assert.assertThrows;
+import static org.junit.Assert.assertTrue;
 
+@RunWith(Parameterized.class)
 public class PluginScannerTest {
 
+    private enum ScannerType { Reflection, ServiceLoader };
+
     @Rule
     public TemporaryFolder pluginDir = new TemporaryFolder();
 
+    public PluginScanner scanner;
+
+    @Parameterized.Parameters
+    public static Collection<Object[]> parameters() {
+        List<Object[]> values = new ArrayList<>();
+        for (ScannerType type : ScannerType.values()) {
+            values.add(new Object[]{type});
+        }
+        return values;
+    }
+
+    public PluginScannerTest(ScannerType scannerType) {
+        switch (scannerType) {
+            case Reflection:
+                this.scanner = new ReflectionScanner();
+                break;
+            case ServiceLoader:
+                this.scanner = new ServiceLoaderScanner();
+                break;
+            default:
+                throw new IllegalArgumentException("Unknown type " + scannerType);
+        }
+    }
+
     @Test
-    public void testLoadingUnloadedPluginClass() {
-        DelegatingClassLoader classLoader = initClassLoader(
+    public void testScanningEmptyPluginPath() {
+        PluginScanResult result = scan(
                 Collections.emptyList()
         );
-        for (String pluginClassName : TestPlugins.pluginClasses()) {
-            assertThrows(ClassNotFoundException.class, () -> classLoader.loadClass(pluginClassName));
-        }
+        assertTrue(result.isEmpty());
     }
 
     @Test
-    public void testLoadingPluginClass() throws ClassNotFoundException {
-        DelegatingClassLoader classLoader = initClassLoader(
+    public void testScanningPluginClasses() {
+        PluginScanResult result = scan(
                 TestPlugins.pluginPath()
         );
+        Set<String> classes = new HashSet<>();
+        result.forEach(pluginDesc -> classes.add(pluginDesc.className()));
         for (String pluginClassName : TestPlugins.pluginClasses()) {
-            assertNotNull(classLoader.loadClass(pluginClassName));
-            assertNotNull(classLoader.pluginClassLoader(pluginClassName));
+            assertTrue("Expected " + pluginClassName + "to be discovered but it was not",
+                    classes.contains(pluginClassName));
```

Comment:
I was wondering about replacing this part with a more rigorous assertion:
```java
Set<String> expectedClasses = new HashSet<>(TestPlugins.pluginClasses());
assertEquals(expectedClasses, classes);
```

But this fails with reflection-based scanning (and succeeds with service loader-based scanning) because the `VersionMethodThrowsConnector` class gets picked up by the plugin scanner, but isn't included in the result of `TestPlugins::pluginClasses`.

If we add a service loader manifest for the `VersionMethodThrowsConnector` class, then both cases fail. And then, if we change `includeByDefault` for the `BAD_PACKAGING_VERSION_METHOD_THROWS_CONNECTOR` to `true`, both test cases pass.

Should we add these changes? They give us slightly better coverage and more helpful assertion messages, and also align the total set of loaded classes when using each type of scanner (which IIUC matches the ideal scenario that we're aiming for once users have finished using the [KIP-898 migration script](https://cwiki.apache.org/confluence/display/KAFKA/KIP-898%3A+Modernize+Connect+plugin+discovery#KIP898:ModernizeConnectplugindiscovery-PluginPathManagementScript)).

Reply:
Thanks, I added both of your suggested fixes. The missing manifest and incorrect includeByDefault were oversights on my part, so the stronger test is a good idea.

Your suggestion also made me realize that I missed some ServiceLoader entries, meaning that ServiceLoaderScanner tests were incomplete. After I added the entries, two tests fail with an error (not an assertion) because the ServiceLoaderScanner doesn't catch the NoClassDefFound (LinkageError) that the ReflectionScanner does.

I added the failing MissingSuperclassConverter and NonExistentInterface test case in KAFKA-14649 #13182, to assert that they no longer crashed the worker, which I thought was reasonable at the time. It appears that the ServiceLoader takes the opinion that LinkageError (like other Errors) should not be caught due to the severity of the cause, which is making me second-think catching it in the ReflectionScanner.

I think there are some options here:
1. Delete the MissingSuperclassConverter manifest to disable the test case for the ServiceLoaderScanner
2. Catch the LinkageError in ServiceLoaderScanner making it shadow other plugins in the same PluginLocation.
3. Make ServiceLoaderScanner behave like ReflectionScanner by forking the ServiceLoader implementation and changing one catch clause.
4. Make ReflectionScanner behave like ServiceLoaderScanner and kill the worker for missing dependencies for plugin classes, restoring the pre-KAFKA-14649 behavior for LinkageError.

Of those options, I think i'm going to implement (4), as i'm loathe to fork ServiceLoader this early in the project, and the severity and breadth of causes for LinkageError means we probably shouldn't handle it.

What do you think? I've pushed the MissingSuperclassConverter manifest so the PluginScannerTest is currently failing.

Reply:
Of the approaches you've laid out, I'm -1 on 1 (the coverage provided by these cases is important, especially if it leads to discoveries like this one) and 2 (if a user runs into this it's going to be a massive headache to debug since they'll see plugin loading errors for more than just the broken plugin).

I agree that it's a bit early to go about implementing our own service loading logic, so although 3 is reasonable in theory, we can hold off on it for now.

I don't love 4, since it's still a step backwards and killing workers on startup is quite extreme. I think I could live with it if we believe there are no better alternatives.

I did some digging and it seems like this is due to a known bug in the OpenJDK `ServiceLoader`: https://bugs.openjdk.org/browse/JDK-8196182 (unclear if this also affects other JDK distributions).

Given that we're dealing with buggy behavior in a popular JDK distribution, it seems reasonable to try deal with this scenario gracefully. I prototyped a local workaround for this that's a little hacky but does cause our tests to pass. We can wrap calls to `Iterator::hasNext` in some error-handling logic:

```java
private boolean serviceLoaderHasNext(Iterator<?> serviceLoaderIterator) {
    try {
        return serviceLoaderIterator.hasNext();
    } catch (LinkageError e1) {
        log.error("Failed to scan for next service loaded plugin", e1);
        try {
            return serviceLoaderIterator.hasNext();
        } catch (LinkageError e2) {
            // It's difficult to know for sure if the iterator was able to advance past the first broken
            // plugin class, or if it will continue to fail on that broken class for any subsequent calls
            // to Iterator::hasNext or Iterator::next
            // As a best-effort measure, we compare exception messages with the assumption that
            // they will include the name of the plugin class; if the messages are the same, then we
            // fail fast and throw the original linkage error; if the messages differ, we try again
            // For reference, see https://bugs.openjdk.org/browse/JDK-8196182, which describes
            // the behavior we are trying to mitigate with this logic as buggy, but indicates that a fix
            // in the JDK standard library ServiceLoader implementation is unlikely to land
            if (Objects.equals(e1.getMessage(), e2.getMessage())) {
                throw e1;
            } else {
                log.error("Failed to scan for next service loaded plugin", e2);
                return true;
            }
        }
    }
}
```

Which would then change our for-loop in `PluginScanner::getServiceLoaderPluginDesc` to this:

```java
Iterator<T> iterator = serviceLoader.iterator();
while (serviceLoaderHasNext(iterator)) {
```

This essentially implements option 4, with some logic to try to intercept linkage errors and handle them gracefully before falling back on terminating the worker if it appears that our service loader is damaged beyond repair.

Thoughts?

Reply:
I looked more into how the Reflections library handles this, and it actually just WARN logs these classes and never shows them to us, so we don't even get the opportunity to log the error ourselves:
```
[2023-07-14 11:54:26,418] WARN could not get type for name test.plugins.MissingSuperclassConverter from any class loader (org.reflections.Reflections:318)
org.reflections.ReflectionsException: could not get type for name test.plugins.MissingSuperclassConverter
	at org.reflections.ReflectionUtils.forName(ReflectionUtils.java:312)
	at org.reflections.ReflectionUtils.lambda$forNames$22(ReflectionUtils.java:330)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1621)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at org.reflections.ReflectionUtils.forNames(ReflectionUtils.java:332)
	at org.reflections.Reflections.getSubTypesOf(Reflections.java:404)
	at org.apache.kafka.connect.runtime.isolation.ReflectionScanner.getPluginDesc(ReflectionScanner.java:118)
	at org.apache.kafka.connect.runtime.isolation.ReflectionScanner.scanPlugins(ReflectionScanner.java:91)
	at org.apache.kafka.connect.runtime.isolation.PluginScanner.scanUrlsAndAddPlugins(PluginScanner.java:78)
	at org.apache.kafka.connect.runtime.isolation.PluginScanner.discoverPlugins(PluginScanner.java:66)
Caused by: java.lang.NoClassDefFoundError: test/plugins/NonExistentInterface
```

I think this takes option 4 off the table, as the reflection-loaded plugins have never failed due to LinkageErrors, before or after KAFKA-14649 change. The service-loaded plugins have always failed due to LinkageErrors.

I really like your implementation; it allows ServiceLoaders which make progress during exceptional hasNext calls to proceed to find other implementations and prevent shadowing. I'll tweak it a little to make it accommodate multiple plugins failing with LinkageErrors before triggering fallback behavior.

And I think I can accept re-throwing the error as a fallback behavior for ServiceLoaderScanner, if it only appears in situations where:
1. The JVM's implementation of ServiceLoader's Iterator's hasNext does not make progress on LinkageError
2. A plugin is installed which throws a LinkageError
3. plugin.discovery is in HYBRID_WARN, HYBRID_FAIL, or SERVICE_LOAD.

The first two conditions should be exceedingly rare, the overall situation has three different possibilities for remediation if it appears in production, and it allows us to still use the upstream ServiceLoader implementation. I'm sold.

Reply:
Oh and just to motivate the tweaks I made to your implementation:
If the ServiceLoader implementation can skip over some LinkageErrors, I think it's possible to have two different plugins throw the same exception and accidentally trigger the equals condition.

For example, if two plugins implemented NonExistentInterface and appeared consecutively in the ServiceLoader manifest, I think the logic should show at least one of the errors, but not fail the worker or shadow other plugins. I think this would typically happen if a plugin packages both a Source and Sink that both have the same missing dependency, but the ServiceLoader implementation could continue to make progress.

There's a heuristic for 100 consecutive failures: if you package 100 consecutive faulty plugins (or have them all on the classpath) then it falls back to failing the worker. I don't think i've seen more than ~5-10 connectors, and ~20-30 Transforms packaged together. If the errors are non-consecutive then the counter resets as well, so the actual number of tolerated hasNext calls could be quite large, covering most use-cases.

The heuristic is just there to prevent infinite loops if the exception message happens to contain a memory address, or the exception alternates between two different messages in the ServiceLoader-not-making-progress case.

Reply:
This all looks great, thanks 👍

One small request for clarification: to me, this implies that we'd fail the worker without trying 100 times to invoke `hasNext`:

> If the exception message happens to contain a memory address, or the exception alternates between two different messages, we should still fail the worker.

But the current implementation still goes through the 100 tries before giving up. I agree with the implementation (better to err on the side of not unnecessarily killing the worker), but just wanted to double check--this is intentional, right?

Also, I notice that we've added `LinkageError` to the `catch` clause around invoking `Iterator::next`. Do we think it's worth adding similar logic to prevent infinite loops in the event that a call to `Iterator::next` that throws an exception also fails to advance the iterator? Or is this unreasonable paranoia?

Reply:
> But the current implementation still goes through the 100 tries before giving up. I agree with the implementation (better to err on the side of not unnecessarily killing the worker), but just wanted to double check--this is intentional, right?

Yes it is intentional. This edit should be more clear:

> The heuristic is just there to prevent infinite loops in the ServiceLoader-not-making-progress case.  If the exception message happens to contain a memory address, or the exception alternates between two different messages, we should still **eventually** fail the worker **and not be stuck in an infinite loop**.

> Also, I notice that we've added LinkageError to the catch clause around invoking Iterator::next. Do we think it's worth adding similar logic to prevent infinite loops in the event that a call to Iterator::next that throws an exception also fails to advance the iterator? Or is this unreasonable paranoia?

I suppose if we're being paranoid about hasNext, being paranoid about next() for JDK8 implementations would also make sense. I've made the error-handling function more generic and applied it to all of the ServiceLoader operations (including load and iterator, which i've never seen fail)